### PR TITLE
DOC Add documentation for packages in pyodide [skip ci]

### PR DESCRIPTION
**Pull request by ryanking13**
Related: #785

Currently, we need to check the git repository manually to see which packages are included in pyodide. This PR adds a script that dynamically generates documentation for a list of packages (name and version) in [pyodide/packages](https://github.com/pyodide/pyodide/tree/main/packages) directory.

Actually, generating a markdown table dynamically in this way seems a little bit ugly to me. So please let me know if there are better ideas.

Rendered version: https://pyodide--1874.org.readthedocs.build/en/1874/usage/packages-in-pyodide.html

### CODE REVIEW

**Comment by ryanking13** [comment]
I can locally build the documentation with `make html`, but readthedocs is not building as expected.

**Comment by hoodmane** [comment]
Can you fix formatting issues (run `make lint` and make sure it passes).
@rth Would be nice if [skip ci] didn't skip format check.

**Comment by ryanking13** [comment]
@hoodmane Formatting issues are fixed in [e52d428](https://github.com/pyodide/pyodide/pull/1874/commits/e52d42809caa866914bde24044db13b981350733). But I will trigger CI for reconfirmation!

File: docs/build_pyodide_packages_list.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+PACKAGES_DIR = base_dir / "packages"
+OUTPUT_FILE = "packages-in-pyodide.md"
+OUTPUT_PATH = base_dir / "docs" / "usage" / OUTPUT_FILE
+
+TEMPLATE = """
+(packages-in-pyodide)=
+# Packages built in Pyodide
+
+The list of prebuilt python packages in pyodide.
```
Comment (rth) [inline docs/build_pyodide_packages_list.py:19]:
```suggestion
The list of prebuilt Python packages in the current version of Pyodide.
```

File: docs/build_pyodide_packages_list.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    is_library = yaml_data.get("build", {}).get("library", False)
+
+    return name, version, is_library
+
+
+def get_package_metadata_list(directory: pathlib.Path):
+    return directory.glob("**/meta.yaml")
+
+
+def to_markdown(template, packages):
```
Comment (rth) [inline docs/build_pyodide_packages_list.py:44]: Could you please add type annotations everywhere, for function input arguments and output? That would help readability particularly in the case where there are no docstrings (which is fine here).

File: docs/usage/loading-packages.md [hunk 3]
```diff
@@ -19,7 +19,7 @@ download all packages that the code snippet imports. This is particularly useful
 for making a repl since users might import unexpected packages. At present,
 {any}`loadPackagesFromImports <pyodide.loadPackagesFromImports>` will not
 download packages from PyPi, it will only download packages included in the
-Pyodide distribution.
+Pyodide distribution. See {ref}`packages-in-pyodide` to check the full list of packages included in pyodide.
```
Comment (rth) [inline docs/usage/loading-packages.md:22]:
```suggestion
Pyodide distribution. See {ref}`packages-in-pyodide` to check the full list of packages included in Pyodide.
```

File: docs/build_pyodide_packages_list.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+OUTPUT_PATH = base_dir / "docs" / "usage" / OUTPUT_FILE
+
+TEMPLATE = """
+(packages-in-pyodide)=
+# Packages built in Pyodide
+
+The list of prebuilt python packages in pyodide.
+These packages can be loaded through {{any}}`pyodide.loadPackage` or {{any}}`micropip.install`.
+See {{ref}}`loading_packages` for information about loading packages.
+
```
Comment (rth) [inline docs/build_pyodide_packages_list.py:22]: Maybe

```suggestion
Note that in addition to this list, pure Python packages with wheels can be loaded
directly from PyPi with {{any}}`micropip.install`.
```

**Review by rth — COMMENTED** [review]
Thanks @ryanking13! This would indeed be very helpful.

> Actually, generating a markdown table dynamically in this way seems a little bit ugly to me. So please let me know if there are better ideas.

Well, we already have a [custom sphinx extension in Pyodide](pyodide/docs/sphinx_pyodide/sphinx_pyodide). So if we wanted to not write that file manually, I imagine we could 
 - copy your scripts there
 - [register a new sphinx directive](https://github.com/pyodide/pyodide/blob/main/docs/sphinx_pyodide/sphinx_pyodide/__init__.py#L77), for instance `pyodide-package-list`
 - [define this directive](https://docutils.sourceforge.io/docs/howto/rst-directives.html)
- then use it as,
```
... pyodide-package-list ::
``` 

But then this would require some effort, so honestly I'm also OK with the current approach. Up to you if you want to try to define a new sphinx directive instead.

A few minor comments, otherwise LGTM.

**Comment by rth** [comment]
> Would be nice if [skip ci] didn't skip format check. 

I know, but since we have no control over what it does, then we are back custom logic as in https://github.com/pyodide/pyodide/pull/1262. Maybe this could have been configured in CI easier in GH Actions, but with CircleCI I have not found a better solution so far.

**Comment by ryanking13** [comment]
@rth Thanks a lot! I didn't know we can make custom directives in Sphinx.

I updated:
- Applied comments
- Add type hints
- Add Sphinx extension for generating package list

**Pull request merged**